### PR TITLE
combine size_average and reduce args in loss functions

### LINKED ISSUE #7929: Combine size_average and reduce args in Loss modules
Currently, our loss modules have two boolean args for size_average and reduce. These flags don't operate independently, so it makes sense to have a single enum (or similar) argument instead. This will also make it easier for us to introduce different averaging behaviors, such as averaging over each individual element and averaging over the size of the batch.

### PR DESCRIPTION
**Pull request by li-roy**
closes #7929

### CODE REVIEW

**Comment by apaszke** [comment]
I can see why we might want to change the C code, but do we really think that breaking backward compatibility to make people say `reduce=Reduce.NONE` instead of `reduce=False` is a good idea? It's certainly less Pythonic. The new thing is much more verbose, and the old API seemed to work quite well.

**Comment by zou3519** [comment]
In Python-land, maybe "reduce=None" should default to `reduce=Reduce.NONE`?

**Comment by li-roy** [comment]
This current version won't break backwards compatibility, but I was planning to stop supporting it in a future version.

A couple motivators behind this change:
- size_average and reduce don't act independently of each other, this seems like bad API design. 
- we want to add more support for other times of reduction, like average by dividing by batch size instead of number of elements. I don't think it's a good idea to keep adding more boolean flags

**Comment by apaszke** [comment]
`reduce` is not a keyword, but a name of a builtin that was actually removed in Python 2 (it's no longer imported by default).

I really don't think that using enums is very user friendly. It's definitely more verbose, and very rare in Python APIs. What's the issue with the two boolean flags we have right now? Sure, there are configurations which are incompatible, but assuming we keep our current defaults, you only specify at most one of those two arguments, and no issues should arise.

**Comment by fmassa** [comment]
I think there is value removing both boolean flags and keeping only a single one.
I'd say that the behavior of `size_average` is unexpected in a few cases, and this could be a way of fixing it without breaking backward compatibility.

I also agree that enum is not very pythonic: what about having strings instead?

**Comment by SsnL** [comment]
I definitely have been confused by what certain combination would do, so at least in the front of making the behavior clearer to users, this certainly helps.

That said, enum indeed isn't really used in Python API, as @apaszke said. A more common pattern is to directly use strings, as we are already doing in `F.pad`.

**Comment by apaszke** [comment]
Strings aren’t ideal, but they’re better than enums. I would be ok with that

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -273,7 +273,7 @@
     CPU: _cosh_out_cpu
     CUDA: _cosh_out_cuda
 
-- func: cosine_embedding_loss(Tensor input1, Tensor input2, Tensor target, double margin=0.0, bool size_average=true, bool reduce=true) -> Tensor
+- func: cosine_embedding_loss(Tensor input1, Tensor input2, Tensor target, double margin=0.0, int64_t reduction=1) -> Tensor
```
Comment (SsnL) [inline aten/src/ATen/native/native_functions.yaml]: oh these magical numbers :/
can you include the enum header in `NativeFunctions.h` and use the enum here?

File: aten/src/ATen/nn.yaml [hunk 2]
```diff
@@ -1,60 +1,60 @@
 # Loss functions
 
-- name: binary_cross_entropy(Tensor self, Tensor target, Tensor weight={}, bool size_average=true, bool reduce=true)
+- name: binary_cross_entropy(Tensor self, Tensor target, Tensor weight={}, int64_t reduction=1)
```
Comment (SsnL) [inline aten/src/ATen/nn.yaml]: nit: magical number

File: aten/src/THCUNN/generic/AbsCriterion.cu [hunk 3]
```diff
@@ -7,13 +7,12 @@ void THNN_(AbsCriterion_updateOutput)(
            THCTensor *input,
            THCTensor *target,
            THCTensor *output,
-           bool sizeAverage,
-           bool reduce)
+           int64_t reduction)
```
Comment (SsnL) [inline aten/src/THCUNN/generic/AbsCriterion.cu]: here and everywhere else in decl: does using `Reduction` instead of `int64_t` as type not work?

File: torch/nn/functional.py [hunk 4]
```diff
@@ -16,6 +16,43 @@
 from . import grad
 
 
+class Reduction:
```
Comment (SsnL) [inline torch/nn/functional.py]: Is this intended to be part of public API?

Comment (SsnL) [reply torch/nn/functional.py]: Please also add comments that tells people that this and the c++ side enums should be in sync

File: torch/nn/functional.py [hunk 5]
```diff
@@ -1396,51 +1449,72 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Tru
         full: whether to compute full loss, i. e. to add the Stirling
             approximation term. Default: ``False``
             :math:`\text{target} * \log(\text{target}) - \text{target} + 0.5 * \log(2 * \pi * \text{target})`.
-        size_average: By default, the losses are averaged over observations for
-            each minibatch. However, if the field :attr:`size_average` is set to ``False``,
-            the losses are instead summed for each minibatch. Default: ``True``
+        size_average (bool, optional): Deprecated (see :attr:`reduction`). By default,
```
Comment (SsnL) [inline torch/nn/functional.py]: here and other docs: do you want to deprecate `size_average` and `reduce`? If so, I have two suggestions:
1. move them below `reduction` in doc
2. raise deprecation warning if people use old api.

**Review by ssnl — COMMENTED** [review]
I added a few general questions and comments to understand the general goal of this PR. After they are resolved I can make a detailed pass.

**Comment by SsnL** [comment]
Please fix tests and rebase too

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (li-roy) [reply aten/src/ATen/native/native_functions.yaml]: I think I would have to changed the code gen to allow that. I will try it out.

File: aten/src/THCUNN/generic/AbsCriterion.cu (hunk 3 above)
Comment (li-roy) [reply aten/src/THCUNN/generic/AbsCriterion.cu]: I tried to use scoped enum at first, it doesn't work for jit and Zach told me it probably wasn't worth the effort to make jit support scoped enum.

File: torch/nn/functional.py (hunk 4 above)
Comment (li-roy) [reply torch/nn/functional.py]: Originally it was going to be public, but I think i'll make it private now that we're using strings.

File: torch/nn/functional.py (hunk 5 above)
Comment (li-roy) [reply torch/nn/functional.py]: OK to 1.

For number 2. I do raise a deprecation warning. It's in get_reduction_constant in the Reduction class. I call it whenever whenever the nn.functional function or nn module is passed size_average or reduce.

Comment (li-roy) [reply torch/nn/functional.py]: I'm actually doing them in this order because I want them to match the actual order of the args.

File: aten/src/THNN/Reduction.h [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#ifndef REDUCE_H
+#define REDUCE_H
+
+namespace Reduction {
+
+enum Reduction { None, ElementwiseMean, Sum, END};
```
Comment (SsnL) [inline aten/src/THNN/Reduction.h]: comments on what's this and link to the python counterparts.

File: torch/nn/functional.py [hunk 7]
```diff
@@ -16,6 +16,43 @@
 from . import grad
 
 
+class Reduction:
+    @staticmethod
+    def to_enum(reduction):
```
Comment (SsnL) [inline torch/nn/functional.py]: It seems unnecessary to split `to_enum` and `get_reduction_constant` into two methods. The functionals will be called a lot (if used), so let's reduce the python code as much as possible by using one method that takes in all three args.

File: aten/src/THCUNN/generic/AbsCriterion.cu (hunk 3 above)
Comment (SsnL) [reply aten/src/THCUNN/generic/AbsCriterion.cu]: okay. Ideally here should have some comments saying `reduction` is an Enum so it will be easier for our future porter of THNN. But there are way too many instances. So idk if I should ask you to make this change.

Comment (li-roy) [reply aten/src/THCUNN/generic/AbsCriterion.cu]: I put a comment in Reduction.h

File: torch/nn/functional.py (hunk 4 above)
Comment (li-roy) [reply torch/nn/functional.py]: Should this remain public since I'm using it in other files?

File: torch/nn/functional.py (hunk 7 above)
Comment (li-roy) [reply torch/nn/functional.py]: I actually need both of these methods. Sometimes I want to turn the old args into the new strings, for example when I'm calling functional from loss.py. Other times I just want to turn old args into the enums.

File: torch/nn/functional.py (hunk 4 above)
Comment (SsnL) [reply torch/nn/functional.py]: @li-roy Please don't. Just prefixing it with `_`. should be fine :)

**Review by ssnl — APPROVED** [review]
Code looks great. Please prefix the enum class with _. Thanks!

**Comment by li-roy** [comment]
@SsnL You're talking about Reduction in nn/functional.py, right? Should that still be prefixed if I'm using it in other files as well?

**Comment by SsnL** [comment]
@li-roy yes, because that's not part of public api. In other words, it shouldn't show up in user's IDE code autocompletion.

**Comment by li-roy** [comment]
@SsnL Oh that makes sense now, thanks for the explanation.

**Comment by SsnL** [comment]
Sorry for now being clear earlier.

**Comment by SsnL** [comment]
btw, if you haven't, you should rebase & import to phabricator.

**Review by soumith — APPROVED** [review]
